common/src/main/java/io/netty/util/AsciiString.java

```diff
@@ -534,23 +534,40 @@ public boolean contentEqualsIgnoreCase(CharSequence string) {
         }
 
         if (string instanceof AsciiString) {
-            AsciiString rhs = (AsciiString) string;
-            for (int i = arrayOffset(), j = rhs.arrayOffset(), end = i + length(); i < end; ++i, ++j) {
-                if (!equalsIgnoreCase(value[i], rhs.value[j])) {
-                    return false;
+            AsciiString other = (AsciiString) string;
+            byte[] value = this.value;
+            if (offset == 0 && other.offset == 0 && length == value.length) {
```

Comment:
length is already checked in line 532

Reply:
This is different: is checking that `length` is the same as the array's one (which is usually true while offsets are 0).
This is used to help the JIT to correlate the loops invariants with the array's boundaries, to fully remove the bound checks. SOmething similar could be achieved just by using PlatformDependent::getByte, altough is not as ideal, given that it is still subject to the JIT decisions related inlining (and that's why I'm not using it here!).

Reply:
@franz1981 thanks for all the details! Learning a lot :)